File: great-search/sol.pas
program sol;

VAR
    caches: array of longint;
    datas: array of array of longint;
    max_itter, total, i, j, data_length, rank, temp2 : integer;
    temp, line, operande: string;

FUNCTION max(a, b: longint): longint;
    BEGIN
        if a > b then
            max := a
        else
            max := b;
    END;

FUNCTION min(a, b: longint): longint;
    BEGIN
        if a < b then
            min := a
        else
            min := b;
    END;

FUNCTION fn(a, b: longint): longint;
    BEGIN
        case operande of
            '*': fn := a * b;
            '+': fn := a + b;
            'max': fn := max(a, b);
            'min': fn := min(a, b);
        end;
    END;

FUNCTION calc(a,b: longint):longint;
    BEGIN
        if (a<0) then
            BEGIN
                calc := a+b;
                exit;
            END;
        calc := a;
    END;


FUNCTION countNegOnes(arr: array of longint): integer;
    VAR
    i, count: integer;
    BEGIN
    count := 0;
        for i := 0 to High(arr) do
            if arr[i] = -1 then
                count := count + 1;
        CountNegOnes := count;
    END;

PROCEDURE sendToNext(currentProcessRank, dest_rank, dataIndex:integer);
    BEGIN
    caches[dest_rank] := datas[currentProcessRank][dataIndex];
    datas[currentProcessRank][dataIndex] := -1;  // assume -1 means no data
    END;

PROCEDURE share(currentProcessRank, dest_rank, dataIndex:integer);
    BEGIN
        datas[dest_rank][dataIndex] := datas[currentProcessRank][dataIndex];
    END;

PROCEDURE solve();
    VAR
        j, rank, right, temp, current_itter: integer;
        bool: boolean;
    BEGIN
    bool := false;
    j := 0;
    current_itter := 0;
    while (not bool)and(current_itter<max_itter) do
        BEGIN
            current_itter := current_itter +1;
            for rank := 0 to total-1 do
                BEGIN
                    right := (rank + 1) mod total;
                    sendToNext(rank, right, calc(rank-j, data_length));
                    if rank <> 0 then
                        BEGIN
                            datas[rank][calc(rank-j-1, data_length)] := fn(caches[rank], datas[rank][calc(rank-j-1, data_length)]);
                        END;
                    if rank = total-1 then
                        BEGIN
                            datas[0][calc(0-j-1, data_length)] := fn(caches[0], datas[0][calc(0-j-1, data_length)]);
                        END;
                    bool := (countNegOnes(datas[rank]) = data_length-1);
                END;
            j += 1;
        END;

    bool := false;
    j := 0;
    while (not bool)and(current_itter<max_itter) do
        BEGIN
            current_itter := current_itter +1;
            for rank := 0 to total-1 do
                BEGIN
                    right := (rank + 1) mod total;
                    temp := calc(rank-j+1, data_length) mod data_length;
                    share(rank, right, temp);
                END;
            bool := (countNegOnes(datas[rank]) = 0);
            j += 1;
        END;
    END;

PROCEDURE printSolution();
    VAR 
        i, j: longint;
    BEGIN
        for i := 0 to total-1 do
            BEGIN
                write(i);
                for j := 0 to High(datas[i]) do
                    write(' ',datas[i][j]);
                writeln();
            END;
        
    END;


BEGIN
    readln(operande);
    readln(max_itter);
    readln(total);

    SetLength(caches, total);
    SetLength(datas, total);
    for j := 0 to total-1 do
        BEGIN
            rank := -1;
            readln(line);
            temp := '';
            for i := 1 to Length(line) do
                if (line[i] = ' ') or (i = Length(line)) then
                    BEGIN
                        if i = Length(line) then 
                            temp := temp + line[i];
                        Val(temp, temp2);
                        if(rank = -1) then
                            BEGIN
                                rank := temp2;
                                SetLength(datas[rank], 0);
                            END
                        else 
                            BEGIN
                                SetLength(datas[rank], Length(datas[rank])+1);
                                datas[rank][High(datas[rank])] := temp2;
                            END;
                        temp := '';
                    END
                else temp := temp + line[i];
            caches[rank] := datas[rank][rank];
        END;
    data_length := Length(datas[0]);
    
    solve();
    printSolution();
END.

File: great-search/sol.js
/* Common Template Starts */
process.stdin.resume()
process.stdin.setEncoding('utf-8')
let inputString = ''
let currentLine = 0

process.stdin.on('data', (inputStdin) => {
  inputString += inputStdin
})

process.stdin.on('end', (_) => {
  inputString = inputString
    .trim()
    .split('\n')
    .map((string) => {
      return string.trim()
    })

  main()
})

function readline() {
  return inputString[currentLine++]
}
/* Common Template Ends */


let caches
let datas
let max_itter, total, data_length
let operande

const fn = (a, b) => {
  return operande == '*'
    ? a * b
    : operande == '+'
    ? a + b
    : operande == 'max'
    ? Math.max(a, b)
    : operande == 'min'
    ? Math.min(a, b)
    : 0
}
const calc = (a, b) => (a < 0 ? a + b : a)

const countNegOnes = (arr) => {
  let counter = 0
  for (element of arr) {
    if (element === -1) {
      counter++
    }
  }
  return counter
}

const sendToNext = (currentProcessRank, dest_rank, dataIndex) => {
  caches[dest_rank] = datas[currentProcessRank][dataIndex]
  datas[currentProcessRank][dataIndex] = -1
}

const share = (currentProcessRank, dest_rank, dataIndex) => {
  datas[dest_rank][dataIndex] = datas[currentProcessRank][dataIndex]
}

const solve = () => {
  let bool
  let j, right, temp, current_itter

  bool = false
  j = 0
  current_itter = 0
  while (!bool && current_itter < max_itter) {
    current_itter++
    for (let rank = 0; rank < total; rank++) {
      right = (rank + 1) % total
      sendToNext(rank, right, calc(rank - j, data_length))
      if (rank != 0) {
        datas[rank][calc(rank - j - 1, data_length)] = fn(
          caches[rank],
          datas[rank][calc(rank - j - 1, data_length)]
        )
      }
      if (rank === total - 1) {
        datas[0][calc(0 - j - 1, data_length)] = fn(
          caches[0],
          datas[0][calc(0 - j - 1, data_length)]
        )
      }
      bool = countNegOnes(datas[rank]) === data_length - 1
    }
    j++
  }

  bool = false
  j = 0
  while (!bool && current_itter < max_itter) {
    current_itter++
    for (let rank = 0; rank < total; rank++) {
      right = (rank + 1) % total
      temp = calc(rank - j + 1, data_length) % data_length
      share(rank, right, temp)
    }
    bool = datas[total - 1].every((e) => e != -1)
    j++
  }
}

const printSolution = () => {
  for (let i = 0; i < total; i++) {
    process.stdout.write(`${i}`)
    for (let j = 0; j < data_length; j++) {
      process.stdout.write(` ${datas[i][j]}`)
    }
    process.stdout.write('\n')
  }
}

function main() {
  operande = readline()
  max_itter = +readline()
  total = +readline()
  caches = new Array(total)
  datas = new Array(total)

  for (let i = 0; i < total; i++) {
    input = readline()
      .split(' ')
      .map(Number)
    datas[input[0]] = input.slice(1)
    caches[input[0]] = datas[input[0]][input[0]]
  }
  data_length = datas[0].length

  solve()
  printSolution()
}
